wolfcrypt/src/port/maxim/max3266x.c

```diff
@@ -0,0 +1,921 @@
+            return status;
+        }
+        return status;
+    }
+
+    int wc_MXC_TPU_SHA_Copy(wc_MXC_Sha* src, wc_MXC_Sha* dst)
+    {
+        dst->msg = NULL;
+        dst->used = 0;
+        dst->size = 0;
+        XMEMCPY(dst->hash, src->hash, sizeof(dst->hash));
+        return EXIT_SUCCESS;
+    }
+
+    void wc_MXC_TPU_SHA_Free(wc_MXC_Sha* hash)
+    {
+        XFREE(hash->msg, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        wc_MXC_TPU_SHA_Init(hash);
+        return EXIT_SUCCESS;
+    }
+
+    int wc_MXC_TPU_SHA_GetDigest(wc_MXC_Sha *hash, unsigned char* digest,
+                                            MXC_TPU_HASH_TYPE algo)
+    {
+        if(hash->msg == 0 && hash->size == 0 && digest != NULL){
+            switch(algo){
+                #ifndef NO_SHA256
+                case MXC_TPU_HASH_SHA256:
+                    XMEMCPY(digest, MXC_EMPTY_DIGEST_SHA256,
+                                            WC_SHA256_DIGEST_SIZE);
+
+                    break;
+                #endif
+                default:
+                    return BAD_FUNC_ARG;
+            }
+            return 0; /* True */
+        }
+        else{
+            return 1; /* False */
+        }
+    }
+
+#if !defined(NO_SHA256)
+
+
+
+    WOLFSSL_API int wc_InitSha256_ex(wc_Sha256* sha256, void* heap, int devId)
+    {
+        if (sha256 == NULL){
+            return 1;
+        }
+        (void)heap;
+        (void)devId;
+        return wc_MXC_TPU_SHA_Init((wc_MXC_Sha *)sha256);
+    }
+
+    WOLFSSL_API int wc_InitSha256(wc_Sha256* sha256)
+    {
+        return wc_InitSha256_ex(sha256, NULL, INVALID_DEVID);
+    }
+
+    WOLFSSL_API int wc_Sha256Update(wc_Sha256* sha256, const unsigned char* data, 
+                                        unsigned int len)
+    {
+        return wc_MXC_TPU_SHA_Update(sha256, data, len);
+    }
+
+    WOLFSSL_API int wc_Sha256Final(wc_Sha256* sha256, unsigned char* hash)
+    {
+        return wc_MXC_TPU_SHA_Final((wc_MXC_Sha *)sha256, hash, 
+                                        MXC_TPU_HASH_SHA256);
+    }
+    
+    WOLFSSL_API int wc_Sha256GetHash(wc_Sha256* sha256, unsigned char* hash)
+    {
+        return wc_MXC_TPU_SHA_GetHash((wc_MXC_Sha *)sha256, hash,
+                                        MXC_TPU_HASH_SHA256);
+    }
+
+    WOLFSSL_API int wc_Sha256Copy(wc_Sha256* src, wc_Sha256* dst)
+    {
+        return wc_MXC_TPU_SHA_Copy((wc_MXC_Sha *)src, (wc_MXC_Sha *)dst);
+    }
+
+    WOLFSSL_API void wc_Sha256Free(wc_Sha256* sha256)
+    {
+        wc_MXC_TPU_SHA_Free((wc_MXC_Sha *)sha256);
+        return;
+    }
+
+#endif
+
+
+
+#endif /* MAX3266X_SHA */
+
+#if defined(MAX3266X_MATH)
+
+    int wc_MXC_MAA_init(unsigned int len)
+    {
+        int status;
+        WOLFSSL_MSG("Setting Hardware Mutex and Starting MAA");
+        status = wolfSSL_CryptHwMutexLock();
+        if (status != EXIT_SUCCESS){
+            return status;
+        }
+        status = MXC_TPU_MAA_Init(len);
```

Comment:
Where is the `wolfSSL_CryptHwMutexUnLock`? Is it the wc_MXC_MAA_Shutdown? Seems like mutex is held much too long.